nucypher/blockchain/eth/signers.py

```diff
@@ -0,0 +1,193 @@
+    def accounts(self) -> List[str]:
+        return NotImplemented
+
+    @abstractmethod
+    def unlock_account(self, account: str, password: str, duration: int = None) -> bytes:
+        return NotImplemented
+
+    @abstractmethod
+    def lock_account(self, account: str) -> str:
+        return NotImplemented
+
+    @abstractmethod
+    def sign_transaction(self, transaction_dict: dict) -> HexBytes:
+        return NotImplemented
+
+    @abstractmethod
+    def sign_message(self, account: str, message: bytes, **kwargs) -> HexBytes:
+        return NotImplemented
+
+
+class Web3Signer(Signer):
+
+    def __init__(self, client: EthereumClient):
+        super().__init__()
+        self.__client = client
+
+    @classmethod
+    def from_signer_uri(cls, uri: str) -> 'Web3Signer':
+        try:
+            blockchain = BlockchainInterfaceFactory.get_or_create_interface(provider_uri=uri)
+        except BlockchainInterface.UnsupportedProvider:
+            raise cls.InvalidSignerURI(f"'{uri}' is not a valid signer URI")
+        signer = cls(client=blockchain.client)
+        return signer
+
+    def is_connected(self) -> bool:
+        return self.__client.w3.isConnected()
+
+    def accounts(self) -> List[str]:
+        return self.__client.accounts
+
+    @validate_checksum_address
+    def is_device(self, account: str):
+        try:
+            # TODO: Temporary fix for #1128 and #1385. It's ugly af, but it works. Move somewhere else?
+            wallets = self.__client.wallets
+        except AttributeError:
+            return False
+        else:
+            HW_WALLET_URL_PREFIXES = ('trezor', 'ledger')
+            hw_accounts = [w['accounts'] for w in wallets if w['url'].startswith(HW_WALLET_URL_PREFIXES)]
+            hw_addresses = [to_checksum_address(account['address']) for sublist in hw_accounts for account in sublist]
+            return account in hw_addresses
+
+    @validate_checksum_address
+    def unlock_account(self, account: str, password: str, duration: int = None):
+        if self.is_device(account=account):
+            unlocked = True
+        else:
+            unlocked = self.__client.unlock_account(account=account, password=password, duration=duration)
+        return unlocked
+
+    @validate_checksum_address
+    def lock_account(self, account: str):
+        if self.is_device(account=account):
+            result = None  # TODO: Force Disconnect Devices?
+        else:
+            result = self.__client.lock_account(account=account)
+        return result
+
+    @validate_checksum_address
+    def sign_message(self, account: str, message: bytes, **kwargs) -> HexBytes:
+        signature = self.__client.sign_message(account=account, message=message)
+        return HexBytes(signature)
+
+    def sign_transaction(self, transaction_dict: dict) -> HexBytes:
+        signed_raw_transaction = self.__client.sign_transaction(transaction_dict=transaction_dict)
+        return signed_raw_transaction
```

Comment:
Why is `account` a parameter of `sign_message()` above and not of `sign_transaction()`? I guess because the signing accounts is already part of the transaction? But then, Clef implementation in L153 does accept an account.

Reply:
That's exactly right, the sender address is included in the transaction dict, implementations may vary slightly, but we can choose to always include it within nucypher.